src/transform/tests/testdata/predicate-pushdown

```diff
@@ -0,0 +1,124 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+cat
+(defsource x [int64 int64])
+----
+ok
+
+# check that equivalences involving runtime constants can be pushed down
+
+build apply=PredicatePushdown
+(join
+  [(get x)
+   (get x)]
+  [[#1 (call_nullary mz_logical_timestamp)]])
+----
+----
+%0 =
+| Get x (u0)
+| Filter (#1 = mz_logical_timestamp())
+
+%1 =
+| Get x (u0)
+
+%2 =
+| Join %0 %1
+| | implementation = Unimplemented
+----
+----
+
+build apply=PredicatePushdown
+(join
+  [(get x)
+   (get x)]
+  [[#1 #3 (call_nullary mz_logical_timestamp)]])
+----
+----
+%0 =
+| Get x (u0)
+| Filter (#1 = mz_logical_timestamp())
+
+%1 =
+| Get x (u0)
+| Filter (#1 = mz_logical_timestamp())
+
+%2 =
+| Join %0 %1
+| | implementation = Unimplemented
+----
+----
+
+# Join equivalence with several runtime constants
+
+build apply=PredicatePushdown
+(join
+  [(get x)
+   (get x)]
+  [[#1 1 (call_nullary mz_logical_timestamp)]])
+----
+----
+%0 =
+| Get x (u0)
+| Filter (#1 = 1), (#1 = mz_logical_timestamp())
+
+%1 =
+| Get x (u0)
+
+%2 =
+| Join %0 %1 (= 1 mz_logical_timestamp())
```

Comment:
Are these two runtime constants supposed to stay in the equivalence class of the join?

Reply:
Good question... I guess it's more correct to lift them as a predicate on top of the join afterwards.

Reply:
Or would it better to just eliminate this equivalence class? 

In the original code, the intent of
```
// For every single-input expression `expr`, we can push
// down `expr = literal` and remove `expr` from the
// equivalence class.
```
is that after this process, the equivalence class will be just `[literal]`. Equivalence classes with less than two expressions then get removed by `canonicalize_equivalences`. 

But in this test case, we are left with two expressions in the equivalence class, so `canonicalize_equivalences` does not know to remove it.

Reply:
Yeah, I think it is safe to assume that if an equivalence class is left only with expressions that don't depend on any input, then it can be removed as an equivalence class.

Reply:
Just pushed a commit to make `canonicalize_equivalences` cover this case.

Reply:
Wouldn't putting the deletion in `canonicalize_equivalences` risk `(= 1 mz_logical_timestamp())` being deleted before predicate pushdown runs? Join fusion and RedundantJoin also call `canonicalize_equivalences`.

Reply:
`(= 1 mz_logical_timestamp())` should not end up as a join equivalence by itself. Only as the result of merging several equality expressions involving input columns, and then extracting pushable predicates from them. Of course, we can create a unit test where we create such equivalence manually, but there is no code path atm that does that.

Reply:
However, since there is no way to prevent that equivalence from being created, I could move the removal into `predicate_pushdown`, and remove it only IFF predicates were pushed down.

Reply:
New commit does that.

However, perhaps what we should do is to push `(1 = mz_logical_timestamp())` to all join inputs instead, to prevent any computation from happening until that condition is true...

Reply:
> (= 1 mz_logical_timestamp()) should not end up as a join equivalence by itself. Only as the result of merging several equality expressions involving input columns, and then extracting pushable predicates from them. Of course, we can create a unit test where we create such equivalence manually, but there is no code path atm that does that.

Good point.

> However, perhaps what we should do is to push (1 = mz_logical_timestamp()) to all join inputs instead, to prevent any computation from happening until that condition is true...

Also a good point.

We may also want to put in an issue to investigate how `1 = mz_logical_timestamp()` should interact with join indexes. Right now, our join implementation planning will want to lift this filter `1 = mz_logical_timestamp()` to use the underlying index, but it is probably significantly cheaper to leave the filter in place.